### PR TITLE
fix Sort by quantity is not working in Magento 2.4 Elastic Search

### LINKED ISSUE #31775: Sort by quantity is not working in Magento 2.4 Elastic Search
The magento 2.4.1 with elastic search is having issues with "Sort by quantity" on the PLP page. The sorting is not working by ASC and DESC after enabling the sorting. 

### Preconditions (*)
1) Magento CE 2.4.1 without sample data is installed.
2) Elastic search is installed 
3) Full-text search for catalog is set to "Elastic Search"

### Steps to reproduce (*)
1) Install magento 2.4.1 with Elastic search installed on the server
2) Log into admin, navigate to "Store-> Product attributes"
3) Find "quantity" attribute and set "Use in filter options" to "Yes"
4) Clear cache and browse the PLP page
5) Sort by quantity

### Expected result (*)
1) The products should get sorted by quantity ASC/DESC based on their stock size

### Actual result (*)
1) Magento sorting is not working based on the quantity
2) The results are remaining the same in the case of ASC and DESC sorting by quantity. The products sorting list will remain the same. 
3) The MySQL query is generating properly with quantity order by param but results are not getting sorted based on available quantity. 

Qty-Asc
Qty-Desc

### Additional info from Engcom
Two products with different quantity
dif_qty
 Sorting by _Quantity_ on the Storefront doesn't change the position of the products
Peek 2021-01-20 17-38

### PR DESCRIPTION
**Pull request by engcom-Charlie**
### Related Pull Requests
https://github.com/magento/inventory/pull/3272

### Fixed Issues (if relevant)

1. Fixes magento/magento2#31775

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by engcom-Charlie** [comment]
@magento run all tests

File: app/code/Magento/Elasticsearch/Elasticsearch5/Model/Adapter/BatchDataMapper/QuantityFieldsProvider.php [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+    /**
+     * Get quantity_and_stock_status fields for data mapper
+     *
+     * @param array $productIds
+     * @param int $storeId
+     * @return array
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function getFields(array $productIds, $storeId)
```
Comment (ihor-sviziev) [inline app/code/Magento/Elasticsearch/Elasticsearch5/Model/Adapter/BatchDataMapper/QuantityFieldsProvider.php:49]: can we define here argument and return  types?
```suggestion
    public function getFields(array $productIds, int $storeId): array
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

File: app/code/Magento/Elasticsearch/Elasticsearch5/Model/Adapter/BatchDataMapper/QuantityFieldsProvider.php (hunk 1 above)
Comment (engcom-Charlie) [reply app/code/Magento/Elasticsearch/Elasticsearch5/Model/Adapter/BatchDataMapper/QuantityFieldsProvider.php:49]: @ihor-sviziev I didn't set argument `storeId` and return types as the same as in the interface.

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Please check failing functional tests

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests CE, Functional Tests EE, Functional Tests B2B

**Comment by engcom-Charlie** [comment]
@magento run all tests

File: app/code/Magento/Elasticsearch/Elasticsearch5/Model/Adapter/BatchDataMapper/QuantityFieldsProvider.php [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    {
+        $fields = [];
+        $attribute = $this->attributeAdapterProvider->getByAttributeCode(self::ATTRIBUTE_CODE);
+        if ($attribute->isSortable()) {
+            foreach ($productIds as $productId) {
+                $fields[$productId][self::ATTRIBUTE_CODE] = $this->stockState->getStockQty($productId);
```
Comment (kandy) [inline app/code/Magento/Elasticsearch/Elasticsearch5/Model/Adapter/BatchDataMapper/QuantityFieldsProvider.php:55]: Request data for each product one by one will be definitely performance problem on big catalogs

**Comment by kandy** [comment]
I have concerns regarding this improvement (it's not a bug as Magento never was designed to have such functionality) at all.
As the quantity of product is constantly changing, the information in the index will be inconsistent
At the same time, the indexer will constantly  eat resources to update the index

File: app/code/Magento/Elasticsearch/Elasticsearch5/Model/Adapter/BatchDataMapper/QuantityFieldsProvider.php (hunk 2 above)
Comment (engcom-Charlie) [reply app/code/Magento/Elasticsearch/Elasticsearch5/Model/Adapter/BatchDataMapper/QuantityFieldsProvider.php:55]: @kandy what if I create separate service to get all quantities by ids with using single query. What do you think?

Comment (kandy) [reply app/code/Magento/Elasticsearch/Elasticsearch5/Model/Adapter/BatchDataMapper/QuantityFieldsProvider.php:55]: Definitely make sense to add additional service (if we don't have it)
but first please pay attention to my top level comments https://github.com/magento/magento2/pull/31990#issuecomment-775325147

**Comment by ddzhura** [comment]
Hi team, this PR requires the Custom Performance Acceptance build to run.
Please, refer to the Panda team on how to run this build before you make a decision on PR Merge.

**Comment by VladimirZaets** [comment]
@magento run Performance Acceptance Tests

**Comment by gabrieldagama** [comment]
https://github.com/magento/magento2/issues/31775#issuecomment-785934470

**Pull request closed without merging**